File: sys/codegen/lib/petscOptionsSetValue/petscOptionsSetValue.c
#include "petscOptionsSetValue.h"
#include "petscOptionsSetValue_types.h"
#include "m2c.h"
#include "petsc4m.h"

static void b_m2c_error(void);

static void c_m2c_error(int varargin_3);

static void m2c_error(void);

static void b_m2c_error(void)
{
  const char *fmt;
  const char *msgid;
  msgid = "Petsc4m:petscOptionsSetValue:InputError";
  fmt = "Argument value must be a null-terminated string.";
  M2C_error(msgid, fmt);
}

static void c_m2c_error(int varargin_3)
{
  const char *fmt;
  const char *msgid;
  msgid = "petsc:RuntimeError";
  fmt = "PetscOptionsSetValue returned error code %d\n";
  M2C_error(msgid, fmt, varargin_3);
}

static void m2c_error(void)
{
  const char *fmt;
  const char *msgid;
  msgid = "Petsc4m:petscOptionsSetValue:InputError";
  fmt = "Argument name must be a null-terminated string.";
  M2C_error(msgid, fmt);
}

void petscOptionsSetValue(const emxArray_char_T *iname,
                          const emxArray_char_T *value, int *errCode,
                          boolean_T *toplevel)
{
  PetscOptions obj;
  const char *iname_data;
  const char *value_data;
  value_data = value->data;
  iname_data = iname->data;
  if ((iname->size[1] != 0) && (iname_data[iname->size[1] - 1] != '\x00')) {
    m2c_error();
  }
  if ((value->size[1] != 0) && (value_data[value->size[1] - 1] != '\x00')) {
    b_m2c_error();
  }
  obj = NULL;
  *errCode = PetscOptionsSetValue(obj, &iname_data[0], &value_data[0]);
  *toplevel = true;
  if (*errCode != 0) {
    c_m2c_error(*errCode);
  }
}

void petscOptionsSetValue_initialize(void)
{
}

void petscOptionsSetValue_terminate(void)
{
}

File: sys/petscOptionsSetValue.m
function [errCode, toplevel] = petscOptionsSetValue(iname, value)
% Sets an option name-value pair in the options database, overriding
% whatever is already present.
%
%   errCode = petscOptionsSetValue(iname, value) sets the option
%   iname in the option database to a string value. The iname must start
%   with '-'. The strings must be null-terminated.
%
%   errCode(int) return code (0 indicates OK)
%
% SEE ALSO: petscOptionsInsertString, petscOptionsHasName petscOptionsGetInt,
%      PetscOptionsGetReal, PetscOptionsGetString
%
% PETSc C interface:
%   PetscErrorCode OptionsSetValue(Mat opts,MatOption op,PetscBool flg)
% http://www.mcs.anl.gov/petsc/petsc-3.7/docs/manualpages/Sys/PetscOptionsSetValue.html

%#codegen -args {coder.typeof(char(0), [1,inf]), coder.typeof(char(0), [1,inf])}

errCode = int32(-1);

if ~isempty(coder.target)
    toplevel = nargout>1;
    if ~isempty(iname) && iname(end) && (toplevel || m2c_debug)
        m2c_error('MPETSc:petscOptionsSetValue:InputError', ...
            'Argument name must be a null-terminated string.')
    end
    if ~isempty(value) && value(end) && (toplevel || m2c_debug)
        m2c_error('MPETSc:petscOptionsSetValue:InputError', ...
            'Argument value must be a null-terminated string.')
    end
    
    errCode = coder.ceval('PetscOptionsSetValue', PETSC_NULL_OPTIONS, ...
        coder.rref(iname), coder.rref(value));
    
    toplevel = nargout>1;
    if errCode && (toplevel || m2c_debug)
        m2c_error('petsc:RuntimeError', 'PetscOptionsSetValue returned error code %d\n', errCode)
    end
end

end
